timemachine/fe/single_topology.py

```diff
@@ -382,141 +359,180 @@ def get_num_connected_components(num_atoms: int, bonds: Collection[Tuple[int, in
+
+        # process chiral volumes, turning off ones at the end-state that have a missing bond.
+
+        # assert presence of bonds
+        if verify:
+            canon_mol_a_bond_idxs_set = {tuple(x) for x in canonicalize_bonds(mol_a_bond_idxs)}
+            for c, i, j, k in mol_a_chiral_atom_idxs:
+                ci = canonicalize_bond((c, i))
+                cj = canonicalize_bond((c, j))
+                ck = canonicalize_bond((c, k))
+                assert ci in canon_mol_a_bond_idxs_set
+                assert cj in canon_mol_a_bond_idxs_set
+                assert ck in canon_mol_a_bond_idxs_set
+
+        mol_c_bond_idxs_set = {tuple(x) for x in mol_c_bond_idxs}
+
+        # Chiral atom restraint c,i,j,k requires that all bonds ci, cj, ck be present at the
+        # end-state in order to be numerically stable under small perturbations due to normalization
+        # along the bond lengths. However, the angle terms defining icj, ick, and jck can be
+        # either 0 or 180, since the normalized chiral volume is still smooth wrt perturbations
+        all_proper_dummy_chiral_atom_idxs_ = []
+        all_proper_dummy_chiral_atom_params_ = []
+        for (c, i, j, k), p in zip(all_dummy_chiral_atom_idxs, all_dummy_chiral_atom_params):
+            if all((c, x) in mol_c_bond_idxs_set or (x, c) in mol_c_bond_idxs_set for x in [i, j, k]):
+                all_proper_dummy_chiral_atom_idxs_.append((c, i, j, k))
+                all_proper_dummy_chiral_atom_params_.append(p)
+            elif verify:
+                warnings.warn(
+                    f"Chiral Volume {c, i, j, k} has a disabled bond, turning off.", ChiralVolumeDisabledWarning
+                )
+
+        all_proper_dummy_chiral_atom_idxs = np.array(all_proper_dummy_chiral_atom_idxs_, np.int32).reshape(-1, 4)
+        all_proper_dummy_chiral_atom_params = np.array(all_proper_dummy_chiral_atom_params_, np.float64)
+
+        mol_c_chiral_atom_idxs = np.concatenate([mol_a_chiral_atom_idxs, all_proper_dummy_chiral_atom_idxs])
+        mol_c_chiral_atom_params = np.concatenate([mol_a_chiral_atom.params, all_proper_dummy_chiral_atom_params])
+
+        # canonicalize bonds
+        mol_c_bond_idxs_canon = canonicalize_bonds(mol_c_bond_idxs)
+        bond_potential = HarmonicBond(mol_c_bond_idxs_canon).bind(np.array(mol_c_bond_params))
+
+        # chiral atoms need special code for canonicalization, since triple product is invariant
+        # under rotational symmetry (but not something like swap symmetry)
+        mol_c_chiral_atom_idxs = canonicalize_chiral_atom_idxs(mol_c_chiral_atom_idxs)
+
+        mol_c_chiral_bond_idxs = canonicalize_bonds(mol_a_chiral_bond_idxs)
+        mol_c_chiral_bond_signs = mol_a_chiral_bond.potential.signs
+
+        chiral_atom_potential = ChiralAtomRestraint(mol_c_chiral_atom_idxs).bind(mol_c_chiral_atom_params)
+        chiral_bond_potential = ChiralBondRestraint(mol_c_chiral_bond_idxs, mol_c_chiral_bond_signs).bind(
```

Comment:
Note: did not end up separating chiral bond restraint setup because the performance impact was negligible compared to the chiral atom restraints